File: download_media.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const url = require('url');

const pagesDir = String.raw`C:\Users\User\Documents\Antigravity\pinelaki-content-backup\pages`;
const mediaDir = String.raw`C:\Users\User\Documents\Antigravity\pinelaki-content-backup\media`;
const mapFile  = String.raw`C:\Users\User\Documents\Antigravity\pinelaki-content-backup\media-map.json`;

fs.mkdirSync(mediaDir, { recursive: true });

const wpPat = /https?:\/\/(?:www\.)?pinelaki\.com\/wp-content\/uploads\/[^\s)"'\]]+/g;

const mediaMap = {};
const allImages = new Map(); // url -> localAbs

const mdFiles = fs.readdirSync(pagesDir).filter(f => f.endsWith('.md'));
console.log(`Processing ${mdFiles.length} pages...`);

for (const fname of mdFiles) {
  const content = fs.readFileSync(path.join(pagesDir, fname), 'utf8');
  const m = content.match(/^url: (.+)/m);
  const pageUrl = m ? m[1].trim() : '';

  const matches = [...content.matchAll(new RegExp(wpPat.source, 'g'))];
  const urls = [...new Set(matches.map(m => m[0].replace(/[)"'\]\s]+$/, '')))];

  const pageImages = [];
  for (const imgUrl of urls) {
    try {
      const parsed = new url.URL(imgUrl);
      const rel = parsed.pathname.replace(/^\//, ''); // wp-content/uploads/...
      const safe = rel.replace(/[\\:*?<>|]/g, '_');
      const localAbs = path.join(mediaDir, ...safe.split('/'));
      pageImages.push({ url: imgUrl, localPath: 'media/' + safe });
      if (!allImages.has(imgUrl)) allImages.set(imgUrl, localAbs);
    } catch (e) {}
  }

  if (pageImages.length && pageUrl) {
    mediaMap[pageUrl] = pageImages;
  }
}

console.log(`Found ${allImages.size} unique images total`);

const headers = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
  'Accept': 'image/webp,image/apng,image/*,*/*;q=0.8',
  'Accept-Language': 'en-US,en;q=0.9',
  'Referer': 'https://pinelaki.com/',
};

function download(imgUrl, dest) {
  return new Promise((resolve) => {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    const parsed = new url.URL(imgUrl);
    const lib = parsed.protocol === 'https:' ? https : http;
    const req = lib.get(imgUrl, { headers, timeout: 30000 }, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        // follow redirect
        return download(res.headers.location, dest).then(resolve);
      }
      if (res.statusCode !== 200) {
        res.resume();
        return resolve({ ok: false, status: res.statusCode });
      }
      const out = fs.createWriteStream(dest);
      res.pipe(out);
      out.on('finish', () => resolve({ ok: true }));
      out.on('error', (e) => resolve({ ok: false, err: e.message }));
    });
    req.on('error', (e) => resolve({ ok: false, err: e.message }));
    req.on('timeout', () => { req.destroy(); resolve({ ok: false, err: 'timeout' }); });
  });
}

async function main() {
  let ok = 0, fail = 0, i = 0;
  const entries = [...allImages.entries()];

  for (const [imgUrl, localAbs] of entries) {
    i++;
    const fname = path.basename(imgUrl);
    const result = await download(imgUrl, localAbs);
    if (result.ok) {
      ok++;
      process.stdout.write(`  [${i}/${entries.length}] OK   ${fname}\n`);
    } else {
      fail++;
      process.stdout.write(`  [${i}/${entries.length}] ERR  ${fname} -- ${result.status || result.err}\n`);
    }
    await new Promise(r => setTimeout(r, 100));
  }

  fs.writeFileSync(mapFile, JSON.stringify(mediaMap, null, 2), 'utf8');

  console.log(`\n=== Done ===`);
  console.log(`Unique images    : ${entries.length}`);
  console.log(`Downloaded OK    : ${ok}`);
  console.log(`Failed           : ${fail}`);
  console.log(`Media map saved  : ${mapFile}`);
}

main().catch(console.error);

File: download_media.py
import os, re, json, urllib.request

pages_dir = r'C:\Users\User\Documents\Antigravity\pinelaki-content-backup\pages'
media_dir = r'C:\Users\User\Documents\Antigravity\pinelaki-content-backup\media'
map_file  = r'C:\Users\User\Documents\Antigravity\pinelaki-content-backup\media-map.json'

os.makedirs(media_dir, exist_ok=True)

wp_pat = re.compile(r'https?://(?:www\.)?pinelaki\.com/wp-content/uploads/[^\s)\]"\']+')

media_map = {}
all_images = {}  # url -> local_abs

md_files = [f for f in os.listdir(pages_dir) if f.endswith('.md')]
print(f'Processing {len(md_files)} pages...')

for fname in sorted(md_files):
    fpath = os.path.join(pages_dir, fname)
    with open(fpath, 'r', encoding='utf-8', errors='replace') as f:
        content = f.read()

    page_url = ''
    m = re.search(r'^url: (.+)', content, re.MULTILINE)
    if m:
        page_url = m.group(1).strip()

    urls = set()
    for m in wp_pat.finditer(content):
        u = m.group(0).rstrip(')]\'"., ')
        urls.add(u)

    page_images = []
    for u in sorted(urls):
        try:
            # Extract path after /wp-content/uploads/
            part = u.split('/wp-content/uploads/', 1)[1]
            part = re.sub(r'[\\:*?<>|]', '_', part)
            local_rel = 'media/wp-content/uploads/' + part
            local_abs = os.path.join(media_dir, 'wp-content', 'uploads', *part.split('/'))
            page_images.append({'url': u, 'localPath': local_rel})
            if u not in all_images:
                all_images[u] = local_abs
        except Exception as e:
            print(f'  SKIP bad URL: {u} ({e})')

    if page_images and page_url:
        media_map[page_url] = page_images

print(f'Found {len(all_images)} unique images')

# Download each image
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Accept': 'image/webp,image/apng,image/*,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.9',
    'Referer': 'https://pinelaki.com/',
}

ok = 0
fail = 0
for i, (url, local_abs) in enumerate(sorted(all_images.items()), 1):
    fname = url.split('/')[-1]
    os.makedirs(os.path.dirname(local_abs), exist_ok=True)
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = resp.read()
        with open(local_abs, 'wb') as f:
            f.write(data)
        ok += 1
        print(f'  [{i}/{len(all_images)}] OK  {fname}')
    except Exception as e:
        fail += 1
        print(f'  [{i}/{len(all_images)}] ERR {fname} -- {e}')

# Save media-map.json
with open(map_file, 'w', encoding='utf-8') as f:
    json.dump(media_map, f, ensure_ascii=False, indent=2)

print(f'\n=== Done ===')
print(f'Pages processed  : {len(md_files)}')
print(f'Unique images    : {len(all_images)}')
print(f'Downloaded OK    : {ok}')
print(f'Failed           : {fail}')
print(f'Media map saved  : {map_file}')
